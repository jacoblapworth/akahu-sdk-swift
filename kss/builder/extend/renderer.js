const { regexDictionary, types } = require('../parser/constants');
const tinycolor = require('tinycolor2');

module.exports = function(handlebars) {
	/**
	 * Extracts the variable properties from the line
	 * returns:
	 *  token name, token value, and optional [ token example ]
	 */
	handlebars.registerHelper('renderTokens', renderTokens);

	/**
	 * Compares the name to see if the container needs to be rendered as inverted.
	 * Similar to the isEqual
	 *
	 * e.g.
	 * {{#isInverted name}}
	 * 		ADD THIS
	 * {{else}} [optional]
	 * 		WILL EXECUTE THIS
	 * {{/isInverted}}
	 */
	handlebars.registerHelper('isInverted', isInverted);
};

const testString = 'To understand what recursion is, you must first understand recursion.';


function renderTokens(variables, block) {
	var newBlock = '';
	variables.split('\n').forEach(line => {
		var test;
		const matches = line.match(/[^:]+/g).map(match => match.trim());

		if (matches[1]) {
			this.token = {};
			this.token.name = matches[0];
			this.token.value = matches[1];
			this.token.title = matches[3];
			this.token.description = matches[4];
			if (matches[2] != undefined) {
				this.token.example = getExample(matches[2], this.token);
			}
			newBlock += block.fn(this);
		}
	});
	return newBlock;
}

function getExample(type, token) {
	switch(type) {
		case types.color:
			const codeTextClass = tinycolor(token.value).isDark() ?
				'ds-colour-code-light'
				: null;
			return `
					<div class="ds-colour-example" style="background-color:${token.value}">
						<code class="${codeTextClass}">${token.value}</code>
						<code class="${codeTextClass}">${token.name}</code>
					</div>`;
		case types.border:
			return `
				<div class="ds-colour-example" style="border:${token.value}">
					<code>${token.name}</code>
				</div>`;
		case types.shadow:
			return `
					<div class="ds-colour-example" style="box-shadow:${token.value}">
						<code>${token.name}</code>
					</div>`;
		case types.fontSize:
			return  `<span class="xui-text-truncated" style="font-size:${token.value};">Aa</span>`;
		case types.fontWeight:
			return `<span style="font-weight:${token.value};">Aa</span>`;
		case types.lineHeight:
			return `<span style="line-height:${token.value}">${testString}</span>`;
		case types.spacing:
			return `<div class="ds-spacing-example" style="height:${token.value};width:${token.value};"></div>`;
		default:
			return '';
	}
}

function isInverted(name, options) {
	return regexDictionary.inverted.test(name) ? options.fn(this) : options.inverse(this);
}
