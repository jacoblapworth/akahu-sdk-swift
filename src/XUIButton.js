import React, {PropTypes} from 'react';
import Classes from 'xui-css-classes';
import cn from 'classnames';
import XUIIcon from 'xui-icon';
import XUILoader from 'xui-loader';

// general helpers
const keys = Object.keys;
const values = x => keys(x).map(k => x[k]);

const ButtonClasses = Classes.Button;

/**
 * String constants
 *
 * @private
 * @type {Object}
 */
const CONSTANTS = {
	TYPES: {
		BUTTON: 'button',
		LINK: 'link'
	},
	ELEMENT_TYPES: {
		BUTTON: 'button',
		LINK: 'a'
	},
	BUTTON_TYPES: {
		SUBMIT: 'submit',
		BUTTON: 'button',
		RESET: 'reset'
	},
	VARIANTS: {
		'primary': ButtonClasses.MAIN,
		'create': ButtonClasses.CREATE,
		'link': ButtonClasses.LINK,
		'negative': ButtonClasses.NEGATIVE,
		'standard': ButtonClasses.STANDARD,
		'unstyled': ''
	},
	SIZES: {
		'small': ButtonClasses.SMALL,
		'full-width': Classes.Utility.FULL_WIDTH,
		'full-width-mobile': ButtonClasses.FULL_WIDTH
	}
};

const propTypes = {

	className: PropTypes.string,
	children: PropTypes.node,
	qaHook: PropTypes.string,

	/** @property {boolean} [isDisabled] Determines if the button is disabled or not. */
	isDisabled: PropTypes.bool,

	/** @property {string} [isExternalLink] If true, sets appropriate `rel` values to prevent new page from having access to `window.opener`. Should be used for links pointing at external sites. **/
	isExternalLink: PropTypes.bool,

	/** @property {boolean} [isLoading] If true, shows a loader inside the button and also disables the button to prevent clicking. Can be used in conjunction with isDisabled (which also provides a disabled class)  */
	isLoading: PropTypes.bool,

	/** @property {boolean} [isGrouped] If this button is part of a parent button group */
	isGrouped: PropTypes.bool,

	/** @property {function} onClick Bind a function to fire when the button is clicked */
	onClick: function (props) {
		if (props.type === 'button' && !(typeof props.onClick === 'function')) {
			throw new Error('Non-link buttons require an onClick function.');
		}
	},

	/** @property {function} [onSecondaryClick] Bind a function to fire when the second button in a split button is clicked */
	onSecondaryClick: function (props) {
		if (props.split && !(typeof props.onSecondaryClick === 'function')) {
			throw new Error('Split buttons require a secondary click handler');
		}
	},

	/** @property {string} [variant='standard'] Determines what the purpose of this button is. `standard`, `primary`, `create`, `negative`, `link` or `unstyled`. */
	variant: PropTypes.oneOf(keys(CONSTANTS.VARIANTS)),

	/** @property {string} [size='default'] Modifier for the size of the button. `small`, `full-width`, or `full-width-layout`. */
	size: PropTypes.oneOf(keys(CONSTANTS.SIZES)),

	/** @property {string} [type='button'] The HTML type of this button. `button`, or `link`. Defaults to `button` */
	type: PropTypes.oneOf(values(CONSTANTS.TYPES)),

	/** @property {string} [buttonType='submit'] The type attribute of this button. `submit`, `button`, or `reset`. Defaults to `submit` */
	buttonType: PropTypes.oneOf(values(CONSTANTS.BUTTON_TYPES)),

	/** @property {string} [href] The `href` attribute to use on the anchor element (ignored unless `type` is `link`) */
	href: function (props) {
		if (props.type === CONSTANTS.LINK && !props.onClick && !props.href) {
			throw new Error('Link buttons without an onClick handler require an href.');
		}
	},

	/** @property {string} [rel] The `rel` attribute to use on the anchor element (ignored unless `type` is `link`) */
	rel: PropTypes.string,

	/** @property {number} [tabIndex=0] The HTML tabIndex property to put on the component */
	tabIndex: PropTypes.number,

	/** @property {string} [target] The `target` attribute to use on the anchor element (ignored unless `type` is `link`) */
	target: PropTypes.string,

	/** @property {string} [title] The `title` attribute for this button */
	title: PropTypes.string,

	/** @property {boolean} [split] Changes the button to a split button. Use `onSecondaryClick` with this to create dropdown experiences */
	split: PropTypes.bool
};

/**
 * Default property values for this component.
 *
 * @public
 */
const defaultProps = {
	buttonType: CONSTANTS.BUTTON_TYPES.SUBMIT,
	tabIndex: 0,
	type: CONSTANTS.ELEMENT_TYPES.BUTTON,
	variant: 'standard'
};

/**
 * Returns a class name for the button depending on the button variant string given. Will return
 * undefined if no matching variant is given.
 *
 * @private
 * @param {string} variant - The button variant
 * @return {string} The variant specific classname
 */
const getVariantClass = variant => {
	const variants = CONSTANTS.VARIANTS;
	return variants.hasOwnProperty(variant) ? variants[variant] : ButtonClasses.STANDARD;
};

/**
 * Returns a classname for the button depending on it's disabled state
 *
 * @private
 * @param {boolean} isDisabled - Whether or not the button is disabled
 * @return {string} The disabled state specific classname
 */
const getDisabledClass = isDisabled => isDisabled ? ButtonClasses.IS_DISABLED : null;

/**
 * Returns a class name for the button depending on the button sizing string given. Will return
 * undefined if no matching size is given.
 *
 * @private
 * @param {string} size - The button size
 * @return {string} The size specific class name
 */
const getSizeClass = size => CONSTANTS.SIZES[size];

/**
 * Returns a class name for the button depending on if it has been set to belong to a group
 *
 * @private
 * @param {boolean} isGrouped - Whether or not the button belongs to a group
 * @return {string} The grouped state specific class name
 */
const getGroupClass = isGrouped => isGrouped ? ButtonClasses.GROUPED : null;

/**
 * Replaces any href of `#` or undefined with `javascript:void(0)`. Else returns the passed href.
 *
 * @private
 * @param {string} href - A given link's href
 * @return {string} The href that will be assigned to a link
 */
const getHref = href => (!href || href === '#') ? 'javascript:void(0)' : href;

/**
 * KeyPress handler which will dispatch a click event when the space bar is pressed.
 *
 * @private
 * @param {KeyboardEvent} event
 * @param {Object} props
 */
function handleSpacebarAsClick(event, props) {
	if (props.isDisabled || props.isLoading) {
		let shouldClick;
		if (event.key) {
			shouldClick = event.key === ' ' || event.key === 'Spacebar';
		} else {
			shouldClick = (event.keyCode || event.which) === 32;
		}

		if (shouldClick) {
			// Clicking the space bar causes scrolling by default.  No bueno for a button.
			event.preventDefault();

			// A native event needs to be dispatched to ensure that all
			// browsers will follow the link.
			// Generate a click event with the latest API, if possible.
			// Use document.createEvent for IE 11.
			let clickEvent;
			if (typeof window.Event === 'function') {
				clickEvent = new MouseEvent('click', {
					bubbles: true,
					cancelable: true,
					view: window,
					detail: 0,
					screenX: 0,
					screenY: 0,
					clientX: 0,
					clientY: 0,
					ctrlKey: false,
					altKey: false,
					shiftKey: false,
					metaKey: false,
					button: 0,
					relatedTarget: null
				});
			} else {
				clickEvent = document.createEvent('MouseEvents');
				// If you're seeing a line through initMouseEvent because WebStorm, read the comment above the if statement
				clickEvent.initMouseEvent('click', true, true, 'window', 0, 0, 0, 0, 0, false, false, false, false, 0, null);
			}
			event.target.dispatchEvent(clickEvent);
		}
	}
}

const setupLinkProps = (props, elementProps) => {
	elementProps.role = 'button';
	elementProps.onKeyPress = function(e) {
		handleSpacebarAsClick(e, props);
	};
	elementProps.href = getHref(props.href);
	elementProps.target = props.target;
	elementProps.rel = props.rel;

	if(props.isExternalLink) {
		elementProps.rel = (elementProps.rel ? elementProps.rel + ' ' : '') + 'external noopener noreferrer'
	}

	if (props.isDisabled || props.isLoading) {
		elementProps['aria-disabled'] = true;
	}
};

export default class XUIButton extends React.Component {
	render () {
		const _this = this;
		const props = _this.props;
		const isLink = props.type === CONSTANTS.TYPES.LINK;
		const ElementType = isLink ? CONSTANTS.ELEMENT_TYPES.LINK : CONSTANTS.ELEMENT_TYPES.BUTTON;
		const variantClass = getVariantClass(props.variant);
		const isSplit = props.split && props.onSecondaryClick;
		const isDisabled = props.isDisabled || props.isLoading;
		const children = props.isLoading ? <XUILoader size="small" defaultLayout={false} className={ButtonClasses.LOADER} /> : props.children;

		const classNames = cn(
			ButtonClasses.BASE,
			props.className,
			variantClass,
			getDisabledClass(props.isDisabled),
			getSizeClass(props.size),
			getGroupClass(props.isGrouped)
		);

		const clickHandler = function() {
			if (isLink && isDisabled) {
				event.preventDefault();
			} else if (!isLink || isLink && props.onClick){
				props.onClick.call(_this, ...arguments);
			}
		};

		const secondaryClickHandler = function() {
			if (isLink && isDisabled) {
				event.preventDefault();
			} else {
				props.onSecondaryClick.call(_this, ...arguments);
			}
		};

		// Standard props for all element types
		const elementProps = {
			title: props.title,
			onClick: clickHandler,
			disabled: isDisabled,
			className: classNames,
			tabIndex: isDisabled ? -1 : props.tabIndex
		};

		// Element type specific props
		if (isLink) {
			setupLinkProps(props, elementProps);
		} else {
			elementProps.type = props.buttonType;
		}

		if (isSplit) {
			elementProps.className = `${elementProps.className} ${ButtonClasses.GROUPED}`;
		}

		let Button = (
			<ElementType {...elementProps} data-automationid={props.qaHook}>
				{children}
			</ElementType>
		);

		if (isSplit) {

			Button = (
				<div className={ButtonClasses.GROUPED}>
					{Button}
					<ElementType className={cn(elementProps.className, ButtonClasses.SPLIT)} onClick={secondaryClickHandler}>
						<XUIIcon icon="caret" className={ButtonClasses.CARET}/>
					</ElementType>
				</div>
			);
		}

		return Button;
	}
}

XUIButton.propTypes = propTypes;
XUIButton.defaultProps = defaultProps;
