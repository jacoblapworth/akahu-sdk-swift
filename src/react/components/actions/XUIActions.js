import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import { ns } from '../helpers/xuiClassNamespace';

const baseClass = `${ns}-actions`;

const XUIActions = ({
  children,
  className,
  hasLayout,
  isLinear,
  primaryAction,
  qaHook,
  secondaryAction,
  tagName,
  ...spreadProps
}) => {
  const classes = cn(
    baseClass,
    className,
    hasLayout && `${baseClass}-layout`,
    isLinear && `${baseClass}-linear`,
  );
  const Tag = tagName;

  const clonedPrimary =
    primaryAction &&
    React.cloneElement(primaryAction, {
      className: cn(`${baseClass}--primary`, primaryAction.props.className),
    });

  const clonedSecondary =
    secondaryAction &&
    React.cloneElement(secondaryAction, {
      className: cn(`${baseClass}--secondary`, secondaryAction.props.className),
    });

  return (
    <Tag {...spreadProps} className={classes} data-automationid={qaHook}>
      {children}
      {clonedPrimary}
      {clonedSecondary}
    </Tag>
  );
};

export default XUIActions;

XUIActions.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  /**
   * Applies standard layout settings. Defaults to true
   */
  hasLayout: PropTypes.bool,
  /**
   * Splits buttons to left and right for a linear flow. Can be combined with hasLayout
   */
  isLinear: PropTypes.bool,
  /**
   * XUIButton to style as primary. Alternately, pass in all the children
   */
  primaryAction: PropTypes.element,
  qaHook: PropTypes.string,
  /**
   * XUIButton to style as secondary
   */
  secondaryAction: PropTypes.element,
  /**
   * Wrapper element tag type, for semantic purposes (eg. panel footers). Defaults to "div"
   */
  tagName: PropTypes.string,
};

XUIActions.defaultProps = {
  hasLayout: true,
  tagName: 'div',
};
