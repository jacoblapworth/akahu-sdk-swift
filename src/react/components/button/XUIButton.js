import React from 'react';
import PropTypes from "prop-types";
import cn from 'classnames';
import XUILoader from '../loader/XUILoader';
import { VariantClassNames, SizeClassNames, ButtonTypes } from './private/constants';

/**
 * Returns true if the button is a borderless variant
 *
 * @private
 * @param {string} variant - The button variant
 * @return {boolean} True if is a borderless button
 */
const isBorderlessVariant = variant => {
	return variant.indexOf('borderless') > -1;
}

/**
 * Returns true if the button is an icon variant
 *
 * @private
 * @param {string} variant - The button variant
 * @return {boolean} True if is an icon button
 */
const isIconVariant = variant => {
	return variant.indexOf('icon') > -1;
}

/**
 * Returns a class name for the button depending on the button variant string given. Will return
 * undefined if no matching variant is given.
 *
 * @private
 * @param {string} variant - The button variant
 * @return {string} The variant specific class name
 */
const getVariantClass = variant => {
	return VariantClassNames.hasOwnProperty(variant) ? VariantClassNames[variant] : 'xui-button-standard';
};

/**
 * Replaces any href of `#` or undefined with `javascript:void(0)`. Else returns the passed href.
 *
 * @private
 * @param {string} href - A given link's href
 * @return {string} The href that will be assigned to a link
 */
const getHref = href => (!href || href === '#') ? 'javascript:void(0)' : href;

/**
 * KeyPress handler which will dispatch a click event when the space bar is pressed.
 *
 * @private
 * @param {KeyboardEvent} event
 * @param {Object} props
 */
function handleSpacebarAsClick(event, props) {
	if (props.isDisabled || props.isLoading) {
		let shouldClick;
		if (event.key) {
			shouldClick = event.key === ' ' || event.key === 'Spacebar';
		} else {
			shouldClick = (event.keyCode || event.which) === 32;
		}

		if (shouldClick) {
			// Clicking the space bar causes scrolling by default.  No bueno for a button.
			event.preventDefault();

			// A native event needs to be dispatched to ensure that all
			// browsers will follow the link.
			// Generate a click event with the latest API, if possible.
			// Use document.createEvent for IE 11.
			let clickEvent;
			if (typeof window.Event === 'function') {
				clickEvent = new MouseEvent('click', {
					bubbles: true,
					cancelable: true,
					view: window,
					detail: 0,
					screenX: 0,
					screenY: 0,
					clientX: 0,
					clientY: 0,
					ctrlKey: false,
					altKey: false,
					shiftKey: false,
					metaKey: false,
					button: 0,
					relatedTarget: null
				});
			} else {
				clickEvent = document.createEvent('MouseEvents');
				// If you're seeing a line through initMouseEvent because WebStorm, read the comment above the if statement
				clickEvent.initMouseEvent('click', true, true, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null);
			}
			event.target.dispatchEvent(clickEvent);
		}
	}
}

/**
 * Links styled as buttons require a specific set of attributes to ensure proper functionality and accessibility.
 * This function ensures that those props will be added to the element.
 *
 * @private
 * @param {Object} props
 * @param {Object} elementProps
 */
const setupLinkProps = (props, elementProps) => {
	// If this is just a plain link styled to be a button, the button role is not needed.  However, if this is a link
	// which is styled to look AND function like a button, then we'll need the role.
	if (!props.href || props.onClick) {
		elementProps.role = 'button';
	}
	elementProps.onKeyPress = function(e) {
		handleSpacebarAsClick(e, props);
	};
	elementProps.href = getHref(props.href);
	elementProps.target = props.target;
	elementProps.rel = props.rel;

	if (props.isExternalLink) {
		elementProps.rel = (elementProps.rel ? elementProps.rel + ' ' : '') + 'external noopener noreferrer'
	}

	if (props.isDisabled || props.isLoading) {
		elementProps['aria-disabled'] = true;
	}
};

/**
 * Attempt to focus the root DOM node of the given component, if the DOM node exists
 *
 * @private
 * @param {XUIButton} button
 */
const focusRootNode = button => button.rootNode != null && button.rootNode.focus();

export default class XUIButton extends React.Component {
	focus() {
		focusRootNode(this);
		// Apparently there are times when calling focus won't actually do it.  I think
		// React's getting in the way, but I'm not sure yet....
		if (this.rootNode !== document.activeElement) {
			setTimeout(focusRootNode, 0, this);
		}
	}

	hasFocus() {
		return this.rootNode == null ? false : this.rootNode.contains(document.activeElement);
	}

	render () {
		const xuiButton = this;
		const {
			type,
			variant,
			onClick,
			isDisabled,
			isLoading,
			children,
			className,
			onKeyDown,
			tabIndex,
			size,
			isGrouped,
			qaHook,
			isExternalLink,
			href,
			target,
			rel,
			isLink,
			isInverted,
			retainLayout,
			minLoaderWidth,
			...spreadProps
		} = xuiButton.props;
		const ElementType = isLink ? 'a' : 'button';
		const variantClass = getVariantClass(variant);
		const buttonDisabled = isDisabled || isLoading;
		let buttonChildren = children;

		const loader = (
			<XUILoader
				key={retainLayout && isLoading ? 'button-loader' : null}
				retainLayout={retainLayout}
				size="small"
				defaultLayout={false}
				className="xui-button--loader" />
		);

		if (retainLayout && isLoading) {
			buttonChildren = [
				<div className="xui-u-hidden-content" key='button-children'>{children}</div>,
				loader
			]
		} else if (isLoading) {
			buttonChildren = loader;
		}

		const buttonClassNames = cn('xui-button', className, variantClass, SizeClassNames[size], {
			'xui-button-is-disabled': isDisabled,
			'xui-button-grouped': isGrouped,
			'xui-button-inverted': isInverted && !isBorderlessVariant(variantClass) && !isIconVariant(variantClass),
			'xui-button-borderless-inverted': isInverted && isBorderlessVariant(variantClass) && !isIconVariant(variantClass),
			'xui-button-min-loader-width': minLoaderWidth
		});

		const clickHandler = function() {
			if (isLink && buttonDisabled) {
				event.preventDefault();
			} else if (onClick) {
				onClick.call(xuiButton, ...arguments);
			}
		};

		// Standard props for all element types
		const elementProps = {
			...spreadProps,
			onClick: clickHandler,
			onKeyDown: buttonDisabled ? null : onKeyDown,
			disabled: buttonDisabled,
			className: buttonClassNames,
			tabIndex: buttonDisabled ? -1 : tabIndex
		};

		// Element type specific props
		if (isLink) {
			setupLinkProps({ href, onClick, isExternalLink, target, rel, isDisabled, isLoading }, elementProps);
		} else {
			elementProps.type = type;
		}

		return (
			<ElementType ref={n => xuiButton.rootNode = n} {...elementProps} data-automationid={qaHook}>
				{buttonChildren}
			</ElementType>
		);
	}
}

XUIButton.propTypes = {
	className: PropTypes.string,
	children: PropTypes.node,
	qaHook: PropTypes.string,

	/** Determines if the button is disabled or not. */
	isDisabled: PropTypes.bool,

	/** If true, sets appropriate `rel` values to prevent new page from having access to `window.opener`. Should be used for links pointing at external sites. */
	isExternalLink: PropTypes.bool,

	/** If true, shows a loader inside the button and also disables the button to prevent clicking. Can be used in conjunction with isDisabled (which also provides a disabled class)  */
	isLoading: PropTypes.bool,

	/** If this button is part of a parent button group */
	isGrouped: PropTypes.bool,

	/** A keydown event handler for the button */
	onKeyDown: PropTypes.func,

	/** Bind a function to fire when the button is clicked */
	onClick: PropTypes.func,

	/** Determines the styling variation to apply: `standard`, `primary`, `create`, `negative`, `link`, 'borderless-standard', 'borderless-primary', 'borderless-create', 'borderless-negative', 'borderless-negative', 'icon', 'icon-large', 'icon-inverted', 'icon-inverted-large' or `unstyled`. */
	variant: PropTypes.oneOf(Object.keys(VariantClassNames)),

	/** Modifier for the size of the button. `small`, `full-width`, or `full-width-layout`. */
	size: PropTypes.oneOf(Object.keys(SizeClassNames)),

	/** Whether or not to render this button using an <a> tag */
	isLink: PropTypes.bool,

	/** The type attribute of this button. `submit`, `button`, or `reset`. */
	type: PropTypes.oneOf(Object.keys(ButtonTypes).map(type => ButtonTypes[type])), // Can't use Object.values without polyfilling for older supported browsers

	/** The `href` attribute to use on the anchor element (ignored unless `isLink` is `true`) */
	href: PropTypes.string,

	/** The `rel` attribute to use on the anchor element (ignored unless `isLink` is `true`) */
	rel: PropTypes.string,

	/** The HTML tabIndex attribute value */
	tabIndex: PropTypes.number,

	/** The `target` attribute to use on the anchor element (ignored unless `isLink` is `true`) */
	target: PropTypes.string,

	/** The `title` attribute */
	title: PropTypes.string,

	/** Applies inverted styling */
	isInverted: PropTypes.bool,

	/** When used with `isLoading` this allows the button to retain children width */
	retainLayout: PropTypes.bool,

	/** Use this to specify a min width on the button, when you want to swap to loading states */
	minLoaderWidth: PropTypes.bool
};

XUIButton.defaultProps = {
	tabIndex: 0,
	type: ButtonTypes.button,
	variant: 'standard',
	isLink: false,
	isDisabled: false,
	isExternalLink: false,
	isGrouped: false,
	isLoading: false,
	retainLayout: true
};
