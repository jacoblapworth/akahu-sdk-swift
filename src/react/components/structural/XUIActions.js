import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import { ns } from '../helpers/xuiClassNamespace';

const baseClass = `${ns}-actions`;

export default class XUIActions extends PureComponent {
	render() {
		const {
			className,
			children,
			hasLayout,
			isLinear,
			primaryAction,
			secondaryAction,
			tagName,
			...spreadProps
		} = this.props;
		const classes = cn(
			baseClass,
			className,
			hasLayout && `${baseClass}-layout`,
			isLinear && `${baseClass}-linear`,
		);
		const Tag = tagName;

		const clonedPrimary = primaryAction && React.cloneElement(primaryAction, {
			className: cn(`${baseClass}--primary`, {
				[`${baseClass}--button-spacing`]: hasLayout && !isLinear,
			}),
		});

		const clonedSecondary = secondaryAction && React.cloneElement(secondaryAction, {
			className: `${baseClass}--secondary`,
		});

		return (
			<Tag {...spreadProps} className={classes}>
				{children}
				{clonedPrimary}
				{clonedSecondary}
			</Tag>
		);
	}
}

XUIActions.propTypes = {
	children: PropTypes.node,
	className: PropTypes.string,
	/**
	 * Applies standard layout settings. Defaults to true
	 */
	hasLayout: PropTypes.bool,
	/**
	 * Splits buttons to left and right for a linear flow. Can be combined with hasLayout
	 */
	isLinear: PropTypes.bool,
	/**
	 * XUIButton to style as primary. Alternately, pass in all the children
	 */
	primaryAction: PropTypes.element,
	/**
	 * XUIButton to style as secondary
	 */
	secondaryAction: PropTypes.element,
	/**
	 * Wrapper element tag type, for semantic purposes (eg. panel footers). Defaults to "div"
	 */
	tagName: PropTypes.string,
};

XUIActions.defaultProps = {
	hasLayout: true,
	tagName: 'div',
};
