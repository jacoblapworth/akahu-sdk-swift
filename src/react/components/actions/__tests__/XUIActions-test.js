import React from 'react';
import XUIButton from '../../button/XUIButton';
import XUIActions from '../XUIActions';
import Enzyme, { mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import renderer from 'react-test-renderer';
import { axe, toHaveNoViolations } from 'jest-axe';

Enzyme.configure({ adapter: new Adapter() });
expect.extend(toHaveNoViolations);

describe('<XUI Actions/>', () => {
  const qaHook = 'qaHook';
  const primary = <XUIButton>Primary</XUIButton>;
  const secondary = <XUIButton>Secondary</XUIButton>;

  it('renders the simplest actions with no extra settings passed', () => {
    const testActions = renderer.create(
      <XUIActions primary={primary} secondary={secondary}>
        Testing ðŸ’©
      </XUIActions>,
    );
    expect(testActions).toMatchSnapshot();
  });
  it('renders extra actions classes that are passed in', () => {
    const wrapper = mount(<XUIActions primary={primary} className="testClass" />);
    const tag = wrapper.find(XUIActions);
    expect(tag.hasClass('testClass')).toEqual(true);
  });
  it('renders actions as a linear type layout', () => {
    const wrapper = mount(<XUIActions primary={primary} secondary={secondary} isLinear={true} />);
    expect(wrapper.find('.xui-actions-linear').length).toBe(1);
  });
  it('renders actions without default layout', () => {
    const wrapper = mount(<XUIActions primary={primary} hasLayout={false} />);
    expect(wrapper.find('.xui-actions-layout').length).toBe(0);
  });
  it('renders actions in a different tag, if supplied', () => {
    const wrapper = mount(<XUIActions primary={primary} tagName="header" />);
    const tag = wrapper.find(XUIActions);
    expect(tag.childAt(0).type()).toEqual('header');
  });
  it('renders actions with automation id when qaHook prop is passed in', () => {
    const wrapper = renderer.create(<XUIActions primary={primary} qaHook={qaHook} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should pass accessibility testing', async () => {
    const component = mount(
      <XUIActions primary={primary} secondary={secondary}>
        Testing
      </XUIActions>,
    );
    const results = await axe(component.html());
    expect(results).toHaveNoViolations();
  });
});
