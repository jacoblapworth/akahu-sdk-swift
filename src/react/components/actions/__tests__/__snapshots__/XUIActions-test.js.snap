// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<XUI Actions/> renders actions with automation id when qaHook prop is passed in 1`] = `
<div
  className="xui-actions xui-actions-layout"
  data-automationid="qaHook"
  primary={
    <XUIButton
      fullWidth="never"
      isDisabled={false}
      isExternalLink={false}
      isGrouped={false}
      isLink={false}
      isLoading={false}
      retainLayout={true}
      tabIndex={0}
      type="button"
      variant="standard"
    >
      Primary
    </XUIButton>
  }
/>
`;

exports[`<XUI Actions/> renders the simplest actions with no extra settings passed 1`] = `
<div
  className="xui-actions xui-actions-layout"
  primary={
    <XUIButton
      fullWidth="never"
      isDisabled={false}
      isExternalLink={false}
      isGrouped={false}
      isLink={false}
      isLoading={false}
      retainLayout={true}
      tabIndex={0}
      type="button"
      variant="standard"
    >
      Primary
    </XUIButton>
  }
  secondary={
    <XUIButton
      fullWidth="never"
      isDisabled={false}
      isExternalLink={false}
      isGrouped={false}
      isLink={false}
      isLoading={false}
      retainLayout={true}
      tabIndex={0}
      type="button"
      variant="standard"
    >
      Secondary
    </XUIButton>
  }
>
  Testing ðŸ’©
</div>
`;
